File: Exercices/TGES/csharp_with_dp/StatsMailObserver.cs
using System;
using System.Net.Mail;
using System.IO;

using TheGreatEmailServer;

namespace Main
{
    public class StatsMailObserver : MailObserver
    {
        private string statsFilename;
        private int receivedCount;
        private int rejectedCount;

        public StatsMailObserver(string statsFilename)
        {
            this.statsFilename = statsFilename;
            receivedCount = 0;
            rejectedCount = 0;
        }

        public void MailReceived(MailMessage mail)
        {
            receivedCount++;
            UpdateStats();
        }

        public void MailRejected(MailMessage mail)
        {
            rejectedCount++;
            UpdateStats();
        }

        public void MailStored(MailMessage mail)
        {
        }

        private void UpdateStats()
        {
            using (StreamWriter file = new StreamWriter(statsFilename))
            {
                file.WriteLine($"Received count: {receivedCount}");
                file.WriteLine($"Rejected count: {rejectedCount}");
                file.WriteLine($"Spam ratio: {rejectedCount * 100 / receivedCount}%");
            }
        }
    }
}

File: Exercices/TGES/ruby/stats_mail_observer.rb
class StatsMailObserver
  def initialize(stats_filename)
    @stats_filename = stats_filename
    @received_count = 0
    @rejected_count = 0
  end
  
  def mail_received(mail)
    @received_count += 1
    update_stats
  end
  
  def mail_rejected(mail)
    @rejected_count += 1
    update_stats
  end

  def mail_stored(mail)
  end

  protected
  
  def update_stats
    File.open(@stats_filename, "w") do |file|
      file.puts "Received count: #{@received_count}"
      file.puts "Rejected count: #{@rejected_count}"
      file.puts "Spam ratio: #{@rejected_count * 100 / @received_count}%"
    end
  end
end
